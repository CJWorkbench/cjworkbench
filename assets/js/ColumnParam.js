// Pick a single column
import React from 'react'
import PropTypes from 'prop-types'


export default class ColumnParam extends React.Component {
  constructor(props) {
    super(props);
    this.state = { colNames: [], selectedCol: props.selectedCol };
    this.onChange = this.onChange.bind(this);
  }

  loadColNames() {
    this.props.getColNames()
      .then(cols => {

        // Always make it possible to select (or show) "(None)"
        var colsPlusNone = ['(None)'].concat(cols);
        this.setState({colNames: colsPlusNone});
      });
  }

  // Load column names when first rendered
  componentDidMount() {
    this.loadColNames();
  }

  // Update our checkboxes when we get new props = new selected column
  // And also column names when workflow revision bumps
  componentWillReceiveProps(nextProps) {
    this.setState({selectedCol: nextProps.selectedCol});
    if (this.props.revision != nextProps.revision) {
      this.loadColNames();
    }
  }

  onChange(evt) {
    var colName = this.state.colNames[evt.target.value];
    if (colName == "(None)") {
      colName = ""; // user should see "please select column" not "no column named (None)"
    }
    this.setState({selectedCol: colName});
    this.props.onChange(colName);
  }

  render() {

    // Select the current column name if any, otherwise (None)
    var idx = this.state.colNames.indexOf(this.state.selectedCol);
    if (idx == -1) {
      idx = 0;
    }

    var itemDivs = this.state.colNames.map( (name, idx) => {
        return <option key={idx} value={idx} className='dropdown-menu-item t-d-gray content-3'>{name}</option>;
    });

    return (
        <select
          className='custom-select dropdown-selector'
          value={idx}
          onChange={this.onChange}
          disabled={this.props.isReadOnly}
        >
          {itemDivs}
        </select>
    );
  }
}

ColumnParam.propTypes = {
  selectedCol:    PropTypes.string.isRequired,
  getColNames:    PropTypes.func.isRequired,
  isReadOnly:     PropTypes.bool.isRequired,
  revision:       PropTypes.number.isRequired,
  onChange:       PropTypes.func.isRequired
};
