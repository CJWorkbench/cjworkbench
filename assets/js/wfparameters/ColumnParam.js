// Pick a single column
import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

export default class ColumnParam extends React.PureComponent {
  static propTypes = {
    name: PropTypes.string.isRequired,
    value: PropTypes.string, // or null
    prompt: PropTypes.string, // default 'Select'
    isReadOnly: PropTypes.bool.isRequired,
    allColumns: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired
    })), // or null if unknown
    onChange: PropTypes.func.isRequired // func(colnameOrNull) => undefined
  }

  onChange = (ev) => {
    this.props.onChange(ev.value || null)
  }

  render() {
    const { allColumns, isReadOnly, prompt, name, value } = this.props
    const isLoading = (allColumns === null)

    let className = 'react-select'
    if (isLoading) className += ' loading'

    // Set dropdown list to 1 option of 'loading' as we wait. When clicked, onChange passes null to callback
    const columnOptions = (allColumns || []).map(column => (
      {
        label: column.name,
        value: column.name
      }
    ))
    const selectedOption = columnOptions.find(c => c.value === value)

    // Keeping classNamePrefix since CSS definitions already exist
    return (
      <Select
        name={name}
        options={columnOptions}
        value={selectedOption}
        isLoading={isLoading}
        className={className}
        classNamePrefix='react-select'
        menuPortalTarget={document.body}
        onChange={this.onChange}
        isClearable={false}
        isDisabled={this.props.isReadOnly}
        placeholder={prompt || 'Select'}
      />
    )
  }
}
