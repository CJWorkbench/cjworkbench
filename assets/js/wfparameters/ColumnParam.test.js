import React from 'react'
import ColumnParam from './ColumnParam'
import { mount } from 'enzyme'

describe('ColumnParam', () => {

  const testcols = ['foo', 'bar', 'baz', 'wow', 'word wrap', 'ok then'];
  const noSelectionText = 'Just pick something, user';

  it('renders with noSelectionText', (done) => {
    let wrapper = mount(
        <ColumnParam
          selectedCol='baz'
          getColNames={() => {return Promise.resolve(testcols)}}
          noSelectionText={noSelectionText}
          isReadOnly={true}
          revision={101}
          onChange={()=>{}}
        />);

    // need to give chance for componentdidMount to run
    setImmediate(() => {
      wrapper.update()

      let state=wrapper.state();
      expect(state.selectedCol).toEqual('baz');
      expect(state.colNames).toEqual([noSelectionText].concat(testcols));

      expect(wrapper).toMatchSnapshot();

      done();
    });
  });

  it('renders without noSelectionText', (done) => {
    let wrapper = mount(
        <ColumnParam
          selectedCol='baz'
          getColNames={() => {return Promise.resolve(testcols)}}
          isReadOnly={true}
          revision={101}
          onChange={()=>{}}
        />);

    // need to give chance for componentdidMount to run
    setImmediate(() => {
      wrapper.update()
      let state=wrapper.state();
      expect(state.selectedCol).toEqual('baz');
      expect(state.colNames).toEqual(['Select'].concat(testcols));

      expect(wrapper).toMatchSnapshot();

      done();
    });
  });
});
