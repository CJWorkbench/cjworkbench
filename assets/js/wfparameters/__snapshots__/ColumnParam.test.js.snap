// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnParam renders with noSelectionText 1`] = `
<ColumnParam
  getColNames={[Function]}
  isReadOnly={true}
  name="column"
  noSelectionText="Just pick something, user"
  onChange={[Function]}
  revision={101}
  selectedCol="baz"
>
  <select
    className="custom-select parameter-base dropdown-selector"
    disabled={true}
    name="column"
    onChange={[Function]}
    value={3}
  >
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="0"
      value={0}
    >
      Just pick something, user
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="1"
      value={1}
    >
      foo
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="2"
      value={2}
    >
      bar
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="3"
      value={3}
    >
      baz
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="4"
      value={4}
    >
      wow
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="5"
      value={5}
    >
      word wrap
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="6"
      value={6}
    >
      ok then
    </option>
  </select>
</ColumnParam>
`;

exports[`ColumnParam renders without noSelectionText 1`] = `
<ColumnParam
  getColNames={[Function]}
  isReadOnly={true}
  name="column"
  onChange={[Function]}
  revision={101}
  selectedCol="baz"
>
  <select
    className="custom-select parameter-base dropdown-selector"
    disabled={true}
    name="column"
    onChange={[Function]}
    value={3}
  >
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="0"
      value={0}
    >
      Select
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="1"
      value={1}
    >
      foo
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="2"
      value={2}
    >
      bar
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="3"
      value={3}
    >
      baz
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="4"
      value={4}
    >
      wow
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="5"
      value={5}
    >
      word wrap
    </option>
    <option
      className="dropdown-menu-item t-d-gray content-3"
      key="6"
      value={6}
    >
      ok then
    </option>
  </select>
</ColumnParam>
`;
